File: Web/packages/kernel/test/telemetry-conformance.test.ts
//
//  telemetry-conformance.test.ts — the SHARED telemetry corpus through the TS kernel
//  (OpenSource/Conformance/telemetry/{scrub,queue}.json). The Kotlin twin (:core
//  TelemetryConformanceTest) and the Swift reference (TelemetryPolicy) run the SAME files, so a
//  redaction rule cannot fire on one renderer and not another — which would be a privacy incident
//  with a platform column — and a crash loop cannot send a count on one platform and ten thousand
//  events on the next.
//
//  Imports the module directly rather than through the barrel: the export line in src/index.ts is
//  a shared file and lands with the coordinator's pass (see the F10 handoff block).
//
//  Missing corpus = loud failure — a silently-skipped conformance suite is how drift starts.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  TELEMETRY_PLACEHOLDERS, TELEMETRY_SCRUB_ORDER, telemetryScrubText, telemetryScrubValue,
  telemetryIsSensitiveKey, telemetryArgShape, telemetryFingerprint, telemetryCollapseMessage,
  telemetrySampled, telemetryBackoffMs, TelemetryQueue,
} from "../src/telemetry.ts";

type TextCase = { name: string; input: string; expect: string };
type KeyCase = { key: string; sensitive: boolean };
type ValueCase = { name: string; key: string; value: string; expect: string };
type FingerprintCase = { name: string; source: string; code: string; message: string | null; expect: string };
type SamplingCase = { fingerprint: string; rate: number; expect: boolean };
type BackoffCase = { attempt: number; expect: number };
type QueueStep = { offer?: string; at?: number; batch?: boolean; ack?: number };
type QueueExpect = { outcome?: string; size?: number; dropped?: number; count?: number; batch?: string[] };
type QueueCase = {
  name: string; capacity: number; windowMs: number; maxBatch: number;
  steps: QueueStep[]; expect: QueueExpect[];
};

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/telemetry");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/telemetry not found");
    dir = parent;
  }
}

function corpus(file: string): { [k: string]: unknown } {
  const doc = JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as { [k: string]: unknown };
  assert.equal(doc["version"], 1, `${file}: version`);
  return doc;
}

test("telemetry: the placeholders and the rule order agree with the corpus", () => {
  const doc = corpus("scrub.json");
  assert.deepEqual(doc["placeholders"], { ...TELEMETRY_PLACEHOLDERS });
  assert.deepEqual(doc["order"], [...TELEMETRY_SCRUB_ORDER]);
});

test("telemetry: every redaction rule agrees with the corpus", () => {
  const cases = corpus("scrub.json")["text"] as TextCase[];
  assert.ok(cases.length > 0, "scrub text corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(telemetryScrubText(testCase.input), testCase.expect, testCase.name);
  }
});

test("telemetry: the non-matches survive untouched", () => {
  const cases = corpus("scrub.json")["survives"] as TextCase[];
  assert.ok(cases.length > 0, "scrub survives corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(testCase.expect, testCase.input, `${testCase.name}: a survivor's expect IS its input`);
    assert.equal(telemetryScrubText(testCase.input), testCase.input, testCase.name);
  }
});

test("telemetry: scrubbing is idempotent — a queued event is never scrubbed twice into mush", () => {
  const doc = corpus("scrub.json");
  for (const testCase of [...(doc["text"] as TextCase[]), ...(doc["survives"] as TextCase[])]) {
    assert.equal(telemetryScrubText(testCase.expect), testCase.expect, testCase.name);
  }
});

test("telemetry: the sensitive-key set agrees with the corpus", () => {
  const cases = corpus("scrub.json")["keys"] as KeyCase[];
  assert.ok(cases.length > 0, "key corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(telemetryIsSensitiveKey(testCase.key), testCase.sensitive, testCase.key);
  }
});

test("telemetry: key/value redaction agrees with the corpus", () => {
  const cases = corpus("scrub.json")["values"] as ValueCase[];
  assert.ok(cases.length > 0, "value corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(telemetryScrubValue(testCase.key, testCase.value), testCase.expect, testCase.name);
  }
});

test("telemetry: an arg shape carries keys and types, never values", () => {
  const shape = telemetryArgShape({ amount: 9.99, token: "sk_live_abc", ok: true, items: [1, 2], note: null });
  assert.deepEqual(shape, { amount: "number", items: "array", note: "null", ok: "boolean", token: "string" });
  assert.equal(JSON.stringify(shape).includes("sk_live_abc"), false, "no value may survive into the shape");
  assert.deepEqual(telemetryArgShape(null), {});
});

test("telemetry: the fingerprint fold agrees with the corpus", () => {
  const cases = corpus("queue.json")["fingerprint"] as FingerprintCase[];
  assert.ok(cases.length > 0, "fingerprint corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(telemetryFingerprint(testCase.source, testCase.code, testCase.message), testCase.expect, testCase.name);
  }
  assert.equal(telemetryCollapseMessage(null), "", "an absent message is an empty tail");
});

test("telemetry: sampling is deterministic and agrees with the corpus", () => {
  const cases = corpus("queue.json")["sampling"] as SamplingCase[];
  assert.ok(cases.length > 0, "sampling corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(telemetrySampled(testCase.fingerprint, testCase.rate), testCase.expect,
                 `${testCase.fingerprint} @ ${testCase.rate}`);
    assert.equal(telemetrySampled(testCase.fingerprint, testCase.rate), testCase.expect, "and again, identically");
  }
});

test("telemetry: the backoff schedule agrees with the corpus", () => {
  const cases = corpus("queue.json")["backoff"] as BackoffCase[];
  assert.ok(cases.length > 0, "backoff corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(telemetryBackoffMs(testCase.attempt), testCase.expect, `attempt ${testCase.attempt}`);
  }
});

test("telemetry: the bounded dedupe queue agrees with the corpus", () => {
  const cases = corpus("queue.json")["queue"] as QueueCase[];
  assert.ok(cases.length > 0, "queue corpus must not be empty");
  for (const testCase of cases) {
    assert.equal(testCase.steps.length, testCase.expect.length, `${testCase.name}: one expectation per step`);
    const queue = new TelemetryQueue(testCase.capacity, testCase.windowMs, testCase.maxBatch);
    testCase.steps.forEach((step, index) => {
      const expected = testCase.expect[index]!;
      const where = `${testCase.name}: step ${index}`;
      if (step.offer !== undefined) {
        const actual = queue.offer(step.offer, step.at as number);
        assert.equal(actual.outcome, expected.outcome, `${where}: outcome`);
        assert.equal(actual.size, expected.size, `${where}: size`);
        assert.equal(actual.dropped, expected.dropped, `${where}: dropped`);
        assert.equal(actual.count, expected.count, `${where}: count`);
      } else if (step.batch === true) {
        assert.deepEqual(queue.batch(), expected.batch, `${where}: batch`);
      } else if (step.ack !== undefined) {
        assert.equal(queue.ack(step.ack).size, expected.size, `${where}: size after ack`);
      } else {
        throw new Error(`${where} names no operation`);
      }
    });
  }
});

test("telemetry: revoked consent clears the queue without counting the drop", () => {
  const queue = new TelemetryQueue(8, 1000, 4);
  queue.offer("a", 0);
  queue.offer("b", 10);
  queue.clear();
  assert.equal(queue.size, 0);
  assert.equal(queue.dropped, 0, "consent revocation is not a delivery failure");
});

File: Engine/Android/core/src/test/kotlin/despia/engine/TelemetryConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * The telemetry conformance runner - executes OpenSource/Conformance/telemetry/{scrub,queue}.json
 * through THIS runtime's TelemetryScrub / TelemetryQueuePolicy / TelemetryQueue
 * (parity/F10-telemetry.md). The TS twin (@despia/kernel telemetry.ts,
 * telemetry-conformance.test.ts) and the Swift reference (TelemetryPolicy) run the SAME files, so
 * a redaction rule cannot fire on one renderer and not another - which would be a privacy incident
 * with a platform column - and a crash loop cannot send a count on one platform and ten thousand
 * events on the next.
 *
 * Missing corpus = loud failure - a silently-skipped conformance suite is how drift starts.
 */
class TelemetryConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/telemetry/$name")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/telemetry/$name not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(name: String): Map<String, Any?> {
        val root = json(corpusFile(name).readText()).foundationValue as? Map<String, Any?>
            ?: error("$name: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "$name version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun cases(name: String, key: String): List<Map<String, Any?>> {
        val list = root(name)[key] as? List<Map<String, Any?>> ?: error("$name: no $key[]")
        assertTrue(list.isNotEmpty(), "$name $key corpus must not be empty")
        return list
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun placeholdersAndOrderAgreeWithCorpus() {
        val doc = root("scrub.json")
        val placeholders = doc["placeholders"] as? Map<String, Any?> ?: error("scrub.json: no placeholders")
        assertEquals(placeholders.mapValues { it.value as String }, TelemetryScrub.PLACEHOLDERS, "placeholders")
        val order = (doc["order"] as? List<Any?>)?.map { it as String } ?: error("scrub.json: no order")
        assertEquals(order, TelemetryScrub.ORDER, "rule order")
    }

    @Test
    fun redactionAgreesWithCorpus() {
        for (case in cases("scrub.json", "text")) {
            val name = case["name"] as? String ?: "<unnamed>"
            assertEquals(case["expect"] as String, TelemetryScrub.text(case["input"] as String), name)
        }
    }

    @Test
    fun nonMatchesSurviveUntouched() {
        for (case in cases("scrub.json", "survives")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val input = case["input"] as String
            assertEquals(input, case["expect"] as String, "$name: a survivor's expect IS its input")
            assertEquals(input, TelemetryScrub.text(input), name)
        }
    }

    @Test
    fun scrubbingIsIdempotent() {
        for (key in listOf("text", "survives")) {
            for (case in cases("scrub.json", key)) {
                val expect = case["expect"] as String
                assertEquals(expect, TelemetryScrub.text(expect), case["name"] as? String ?: "<unnamed>")
            }
        }
    }

    @Test
    fun sensitiveKeysAgreeWithCorpus() {
        for (case in cases("scrub.json", "keys")) {
            val key = case["key"] as String
            assertEquals(case["sensitive"] as Boolean, TelemetryScrub.isSensitiveKey(key), key)
        }
    }

    @Test
    fun keyValueRedactionAgreesWithCorpus() {
        for (case in cases("scrub.json", "values")) {
            val name = case["name"] as? String ?: "<unnamed>"
            assertEquals(
                case["expect"] as String,
                TelemetryScrub.value(case["key"] as String, case["value"] as String),
                name,
            )
        }
    }

    @Test
    fun argShapeCarriesTypesNeverValues() {
        val shape = TelemetryScrub.argShape(
            mapOf("amount" to 9.99, "token" to "sk_live_abc", "ok" to true, "items" to listOf(1, 2), "note" to null),
        )
        assertEquals(
            mapOf("amount" to "number", "items" to "array", "note" to "null", "ok" to "boolean", "token" to "string"),
            shape,
        )
        assertFalse(shape.toString().contains("sk_live_abc"), "no value may survive into the shape")
        assertEquals(emptyMap<String, String>(), TelemetryScrub.argShape(null))
    }

    @Test
    fun fingerprintFoldAgreesWithCorpus() {
        for (case in cases("queue.json", "fingerprint")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val actual = TelemetryQueuePolicy.fingerprint(
                case["source"] as String, case["code"] as String, case["message"] as? String,
            )
            assertEquals(case["expect"] as String, actual, name)
        }
        assertEquals("", TelemetryQueuePolicy.collapseMessage(null), "an absent message is an empty tail")
    }

    @Test
    fun samplingIsDeterministicAndAgreesWithCorpus() {
        for (case in cases("queue.json", "sampling")) {
            val fingerprint = case["fingerprint"] as String
            val rate = (case["rate"] as Number).toDouble()
            val expect = case["expect"] as Boolean
            assertEquals(expect, TelemetryQueuePolicy.sampled(fingerprint, rate), "$fingerprint @ $rate")
            assertEquals(expect, TelemetryQueuePolicy.sampled(fingerprint, rate), "and again, identically")
        }
    }

    @Test
    fun backoffScheduleAgreesWithCorpus() {
        for (case in cases("queue.json", "backoff")) {
            val attempt = (case["attempt"] as Number).toInt()
            assertEquals((case["expect"] as Number).toLong(), TelemetryQueuePolicy.backoffMs(attempt), "attempt $attempt")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun boundedDedupeQueueAgreesWithCorpus() {
        for (case in cases("queue.json", "queue")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val steps = case["steps"] as? List<Map<String, Any?>> ?: error("$name: no steps")
            val expect = case["expect"] as? List<Map<String, Any?>> ?: error("$name: no expect")
            assertEquals(steps.size, expect.size, "$name: one expectation per step")

            val queue = TelemetryQueue(
                (case["capacity"] as Number).toInt(),
                (case["windowMs"] as Number).toLong(),
                (case["maxBatch"] as Number).toInt(),
            )
            steps.forEachIndexed { index, step ->
                val want = expect[index]
                val where = "$name: step $index"
                when {
                    step["offer"] != null -> {
                        val actual = queue.offer(step["offer"] as String, (step["at"] as Number).toLong())
                        assertEquals(want["outcome"] as String, actual.outcome.wire, "$where: outcome")
                        assertEquals((want["size"] as Number).toInt(), actual.size, "$where: size")
                        assertEquals((want["dropped"] as Number).toInt(), actual.dropped, "$where: dropped")
                        assertEquals((want["count"] as Number).toInt(), actual.count, "$where: count")
                    }
                    step["batch"] == true -> {
                        val wanted = (want["batch"] as List<Any?>).map { it as String }
                        assertEquals(wanted, queue.batch(), "$where: batch")
                    }
                    step["ack"] != null -> {
                        val size = queue.ack((step["ack"] as Number).toInt())
                        assertEquals((want["size"] as Number).toInt(), size, "$where: size after ack")
                    }
                    else -> error("$where names no operation")
                }
            }
        }
    }

    @Test
    fun revokedConsentClearsTheQueueWithoutCountingTheDrop() {
        val queue = TelemetryQueue(8, 1000L, 4)
        queue.offer("a", 0L)
        queue.offer("b", 10L)
        queue.clear()
        assertEquals(0, queue.size)
        assertEquals(0, queue.dropped, "consent revocation is not a delivery failure")
    }
}
